.github/workflows/workflow.yml

```diff
@@ -0,0 +1,142 @@
+name: Colcon Build, Test, Coverage and Autoformat
+
+on:
+  pull_request:
+    branches:
+      - dev
+      - pre-release
+      - release
+
+  workflow_dispatch:
+
+jobs:
+  Autoformat:
+    runs-on: self-hosted
+    steps:
+      - uses: actions/checkout@main
+        with:
+          ref: ${{ github.head_ref }}
+          fetch-depth: 2
+
+      - name: Clang-format style check
+        uses: THI-Drone/action-clang-format@main
+
+      - name: Autopep8 style check
+        uses: THI-Drone/action-autopep@main
+
+
+  Build-and_test:
+    needs: Autoformat
+    runs-on: self-hosted
+
+    steps:
+
+    - name: Get the date
+      id: date
+      run: echo "date=$(date +'%F_%H-%M-%S')" >> $GITHUB_OUTPUT
+
+    - name: Setup Docker Container 
+      run: | 
+        CONTAINER_NAME="ros-container"
+        IMAGE="ros:humble"
+        OLD_ID=$(docker inspect --format "{{.Id}}" $IMAGE)
+        docker pull $IMAGE
+        NEW_ID=$(docker inspect --format "{{.Id}}" $IMAGE)
+        INSTALL_DEPENDENCIES=false
+        
+        if [ "$OLD_ID" != "$NEW_ID" ]; then
+          echo "update container"
+          docker stop $CONTAINER_NAME || true
+          docker rm $CONTAINER_NAME || true
+          docker run --name $CONTAINER_NAME -i -d $IMAGE bash
+          INSTALL_DEPENDENCIES=true
+        else
+          echo "use old container"
+          CONTAINER_ID=$(docker ps -a -q --filter name="^/${CONTAINER_NAME}$")
+
+          if [ -z "$CONTAINER_ID" ]; then
+            echo "run new container"
+            docker run --name $CONTAINER_NAME  -i -d $IMAGE bash
+            INSTALL_DEPENDENCIES=true
+          else
+            RUNNING_CONTAINER_ID=$(docker ps -q --filter name="^/${CONTAINER_NAME}$")
+
+            if [ -z "$RUNNING_CONTAINER_ID" ]; then
+                docker start $CONTAINER_NAME
+                docker exec -i $CONTAINER_NAME /bin/bash -c "rm -rf /thi-drone-ws/"
+            else
+                echo "Container $CONTAINER_NAME is already running."
+                docker exec -i $CONTAINER_NAME /bin/bash -c "rm -rf /thi-drone-ws/"
+            fi
+          fi
+        fi
+        if [ "$INSTALL_DEPENDENCIES" = true ]; then
+          docker exec -i $CONTAINER_NAME /bin/bash -c "apt-get update && apt-get install -y nlohmann-json3-dev && apt-get install -y python3-pip && pip install setuptools==58.2.0 && pip install pytest && rosdep update && pip install coverage && pip install colcon-coveragepy-result && pip install pytest-cov && apt-get install ros-humble-ament-cmake -y && apt-get install ros-humble-ament-cmake-clang-format -y && apt-get install ros-humble-ament-cmake-cppcheck -y && apt-get install ros-humble-ament-cmake-cpplint -y && apt-get install ros-humble-ament-cmake-flake8 -y && apt-get install ros-humble-ament-cmake-lint-cmake -y && apt-get install ros-humble-ament-cmake-pep257 -y && apt-get install gcc lcov -y && apt-get install python3-colcon-lcov-result && apt-get install lcov -y"
```

Comment:
Wäre es möglich, unser bereits bestehendes Dockerfile als Basis für den Container herzunehmen? Also dass das jedes Mal heruntergeladen und zum Bauen des Containers genutzt wird? Möchte man in der aktuellen Version eine Dependency hinzufügen, muss man in jedes einzelne Repo gehen und das ändern...

Reply:
Das ist nicht der Fall: https://github.com/THI-Drone/thi-drone-ws/blob/dev/.devcontainer/Dockerfile

Die genannten VSCode Sachen werden erst durch VSCode nachinstalliert. Sie sind nicht im Dockerfile enthalten.

Reply:
![grafik](https://github.com/THI-Drone/demo_package/assets/115873997/a3c35e6c-3c75-440f-9a96-23162dc5aabc)


wäre jetzt sogar fertig. wir müssen dem runner nur die berechtigungen geben.
Wir brauchen einen PAT der auf das repo zugreifen darf. Ich kann leider keinen erstellen der org rechte hat. eventuell kann das nur johan ich weiß es nicht.

Reply:
Was willst du denn da machen?

Reply:
hab nen repo erstellt was den workflow automatisch in alle submodule pushen kann. Fehlt nurnoch ein access token. Warum auch immer darf ich in meinem Github account kein personal access token für diese org erstellen

Reply:
Willst du das über das distribute-workflow Repo machen?

Reply:
Und wo ist denn der aktuelle Workflow, den du verteilen willst?

Reply:
Führe bitte nichts aus, solange nicht alles geklärt ist. Ich muss mir das erst genau anschauen.

Reply:
das kann ja eh noch nichts machen weil der token fehlt. Aber du kannst es dir gerne anschauen.

Reply:
wenn du eine bessere idee hast wie man solche änderungen an z.b. configurationsdatieen in allen submodulen oder dem workflow hast dann kannst du es gerne mitteilen. Aber jedesmal in allen submodulen nen pullrequest zu erstellen und den dann von 2 leuten reviewen zu lassen finde ich keine praktikable lösung

Reply:
Würdest du mir bitte meine Frage beantworten?

Reply:
Habe doch auf alles geantwortet außer auf die frage wo der workflow ist.

Der jetzt aktuelle workflow ist in diesem PR damit ihr den reviewen könnt, oder was meinst du? Den kann man wenn wir das so machen wollen dann reviewen und mittels dem distribute-workflow an alle repos verteilen. Wenn nicht dann machen wir halt in jedem repo einen pullrequest auf, sehe ich aber als nicht praktikabel an.

Reply:
> Habe doch auf alles geantwortet außer auf die frage wo der workflow ist.

Genau das war meine Frage 😉.

D.h. der jetzige Plan ist, dass das distribute_workflow Repo die aktuelle Version vom demo_package nimmt und sie in die anderen Repos kopiert?

Reply:
Ich hätte gerne nach wie vor, dass das Dockerfile als Basis genutzt wird und wir nicht noch einen Nebenschauplatz aufbauen: https://github.com/THI-Drone/thi-drone-ws/blob/dev/.devcontainer/Dockerfile

Reply:
ne Plan ist man macht einen PR in einem C package und in einem Python package, damit man sicher gehen kann das es funktioniert. Wenn das 2 leute reviewed haben löscht man die PRs kopiert die datei in das distribute_workflow repo und das verteils die an alle andern repos.

Alternativ macht man halt in jedem repo einen PR auf und reviewed es jedes mal, aber das ist eigentlich nicht notwendig, da ja in jedem repo der gleiche workflow drin ist.

Ist mir eigentlich egal wie wir es machen ich denke nur der aufwand es in jedem repo zu machen ist bisschen unnötig.

Reply:
man könnte es auch aus dem PR raus ziehen, aber dann muss man halt den workflow manuell triggern. Wenn man einfach copy paste macht läuft der workflow automatisch beim commit auf den master branch des distribute_workflow repos

Reply:
das können wir gerne machen, ich habe allerdings jetzt erstmal keine Zeit das zu implementieren. Dennoch würdeich die andere struktur gerne so implementieren, da wenn man beim workflow irgendetwas hinzufügen möchte muss man es sonst dennoch über diesen umweg machen.

Reply:
> Dennoch würdeich die andere struktur gerne so implementieren, da wenn man beim workflow irgendetwas hinzufügen möchte muss man es sonst dennoch über diesen umweg machen.

Den Satz verstehe ich leider nicht.

Reply:
wie schon erläutert. Wenn man etwas am Workflwo ändern muss, dann muss man es in jedem einzelnen repo anpassen, da der workflow in jedem submodul der gleiche ist und auch bleiben soll. Wenn ich jetzt eine änderung (am Workflow) habe muss ich in jedem repo einen pullrequest mit anschließendem review aufmachen, obwohl ein review in demo_package und demo_package_py langt, da der workflow ja in allen packages der gleiche ist und wir nur testen müssen ob alles für Python und C/C++ funktioniert.

Das repo distribute_workflow könnte dann nachdem man den workflow in den beiden genannten repos reviewed und getestst hat einem die arbeit, in allen submodulen einen PR zu erstellen, abnehmen. Das ist unabhängig von dem Docker image und deswegen würde ich dies, sofern die verantwortlichen einverstanden sind so gerne verwenden. Implementiert ist dies schon es müsste nurnoch der key hinzugefügt werden, sodass man es in Betrieb nehmen kann. Wenn du dir anschauen willst wie dies implementiert ist dann schaue im distribute_workflow repo in die datei .github/workflows/workflow.yml.

Dass ich jetzt ein anderes Docker image verwende als das des dev containers liegt an dem fakt, dass ich das mit Jonathan so abgesprochen habe. Mag sein, dass es jetzt nicht optimal ist, aber fürs erste sollte der workflow so funktionieren.

Sobald ich zeit habe werde ich das auch gerne noch ändern, nur habe ich eben vermutlich Heute keine Zeit mehr dazu. Dennoch halte ich es für sinnvoll den workflow, sofern nicht noch Fehler vorhanden sind so schonmal in Betrieb zu nehmen um das autoformatting und das durchlaufen kommender test zu gewährleisten.

Reply:
Vielen Dank für deine ausführliche Erklärung 👍.

Ich bin ehrlich gesagt nicht überzeugt davon, die Änderungen über den Workflow direkt in dev Branch zu (aktuell: force) pushen. Ich sehe da die Gefahr, dass das Dinge kaputt machen bzw. überschreiben könnte.
Mir wäre es lieber, die Skripte würden in einen eigenen Branch gepusht und dann manuell vom Repo-Owner in den aktuellen Feature-Branch gemerged (und dabei auch ausprobiert).
Einen PR nur dafür aufzumachen (mit 2 Reviewern) halte ich auch für übertrieben.

Reply:
Alles klar, ich denke dann können wir so verfahren, dass wir das Thema 2 noch in den Hintergrund stellen. Das einzige was wir jetzt noch klären müssen ist, ob wir ohne die die Änderung an dem Dockerfile schon mergen wollen. Ich bin aktuell der Meinung, dass wir noch warten bis @MatixLoc genügend Zeit hat um das umzubauen und das dann einmal manuell in alle repos reinpacken. Was meint ihr dazu @MatixLoc @Philipp66904 ?

Reply:
Mir wäre es auch lieber, auf die Dockerfile Integration (sofern möglich) zu warten.

Reply:
Das wäre natürlich ideal 👍

Reply:
jo das wäre das beste. Dann erstellen wir einfach ein Repo in dem das liegt und benutzen in allen andern repos nur ein aufruf dieses workflows.

Reply:
Wenn das geht wäre das die beste Strategie um die workflows in die repos zu bekommen.

Reply:
ja offensichtlich geht das. war lost hatte mir das schonmal angesehen und hatte gelesen, dass es im gleichen repo sein muss. In der offiziellen Doku (die du verlinkt hast) steht aber drin dass die auch in einem öffentlichen oder einem repo der gleichen organisation mit gewährten rechten liegen kann.

Sollte also möglich sein

Reply:
Ist dieser PR dann damit nicht mehr relevant oder? Wenn man die action auslagern kann wäre schon deutlich cleaner.